# Income Prediction using ML

In [112]:
import pandas as pd
import numpy as np
import os 
import matplotlib.pyplot as plt
import seaborn as sns

##  Building DataFrame

Data Set Used:
* The "census" data set that contains Census information from 1994: `censusData.csv`


#### Load a Data Set and Save it as a Pandas DataFrame



 I inspected  each the datasets and choose the Census data because I found it more intersting and aligns with my couristy of the Machine Learning problem and question I am trying to solve. There are 15 coloums in this data that I can work with. 


In [ ]:
# File name of my dataset

adultDataSet_filename = os.path.join(os.getcwd(), "data", "censusData.csv")



In [114]:
#fist I want to read and inspect all of the data so I can pick which data set I want to work with
df_census = pd.read_csv(adultDataSet_filename)
df_airbnb = pd.read_csv(airbnbDataSet_filename)
df_whr = pd.read_csv(WHRDataSet_filename)
df_books = pd.read_csv(bookReviewDataSet_filename)



In [115]:
#prints the first 15 for me to inspect and choose 
print("Census")
display(df_census.head(50)) # I choose to work with the censusData so I wanted to read more columns and rows 

print("Airbnb")
display(df_airbnb.head(15))

print("World Happiness")
display(df_whr.head(15))

print("Book Reviews")
display(df_books.head(15))



Census


age         workclass  fnlwgt     education  education-num  \
0   39.0         State-gov   77516     Bachelors             13   
1   50.0  Self-emp-not-inc   83311     Bachelors             13   
2   38.0           Private  215646       HS-grad              9   
3   53.0           Private  234721          11th              7   
4   28.0           Private  338409     Bachelors             13   
5   37.0           Private  284582       Masters             14   
6   49.0           Private  160187           9th              5   
7   52.0  Self-emp-not-inc  209642       HS-grad              9   
8   31.0           Private   45781       Masters             14   
9   42.0           Private  159449     Bachelors             13   
10  37.0           Private  280464  Some-college             10   
11  30.0         State-gov  141297     Bachelors             13   
12  23.0           Private  122272     Bachelors             13   
13  32.0           Private  205019    Assoc-acdm             12   
14  40.0           Private  121772     Assoc-voc             11   
15  34.0           Private  245487       7th-8th              4   
16  25.0  Self-emp-not-inc  176756       HS-grad              9   
17  32.0           Private  186824       HS-grad              9   
18  38.0           Private   28887          11th              7   
19  43.0  Self-emp-not-inc  292175       Masters             14   
20  40.0           Private  193524     Doctorate             16   
21  54.0           Private  302146       HS-grad              9   
22  35.0       Federal-gov   76845           9th              5   
23  43.0           Private  117037          11th              7   
24  59.0           Private  109015       HS-grad              9   
25  56.0         Local-gov  216851     Bachelors             13   
26  19.0           Private  168294       HS-grad              9   
27  54.0               NaN  180211  Some-college             10   
28  39.0           Private  367260       HS-grad              9   
29  49.0           Private  193366       HS-grad              9   
30  23.0         Local-gov  190709    Assoc-acdm             12   
31  20.0           Private  266015  Some-college             10   
32  45.0           Private  386940     Bachelors             13   
33  30.0       Federal-gov   59951  Some-college             10   
34  22.0         State-gov  311512  Some-college             10   
35  48.0           Private  242406          11th              7   
36  21.0           Private  197200  Some-college             10   
37  19.0           Private  544091       HS-grad              9   
38  31.0           Private   84154  Some-college             10   
39  48.0  Self-emp-not-inc  265477    Assoc-acdm             12   
40  31.0           Private  507875           9th              5   
41  53.0  Self-emp-not-inc   88506     Bachelors             13   
42  24.0           Private  172987     Bachelors             13   
43  49.0           Private   94638       HS-grad              9   
44  25.0           Private  289980       HS-grad              9   
45  57.0       Federal-gov  337895     Bachelors             13   
46  53.0           Private  144361       HS-grad              9   
47  44.0           Private  128354       Masters             14   
48  41.0         State-gov  101603     Assoc-voc             11   
49  29.0           Private  271466     Assoc-voc             11   

           marital-status         occupation   relationship  \
0           Never-married       Adm-clerical  Not-in-family   
1      Married-civ-spouse    Exec-managerial        Husband   
2                Divorced  Handlers-cleaners  Not-in-family   
3      Married-civ-spouse  Handlers-cleaners        Husband   
4      Married-civ-spouse     Prof-specialty           Wife   
5      Married-civ-spouse    Exec-managerial           Wife   
6   Married-spouse-absent      Other-service  Not-in-family   
7      Married-civ-spouse    Exec-managerial        Husband   
8           Never-ma

Airbnb


name  \
0                               Skylit Midtown Castle   
1   Whole flr w/private bdrm, bath & kitchen(pls r...   
2            Spacious Brooklyn Duplex, Patio + Garden   
3                    Large Furnished Room Near B'way　   
4                  Cozy Clean Guest Room - Family Apt   
5      Lovely Room 1, Garden, Best Area, Legal rental   
6                    Only 2 stops to Manhattan studio   
7                             UES Beautiful Blue Room   
8   Amazing location! Wburg. Large, bright & tranquil   
9         Perfect for Your Parents:  Privacy + Garden   
10                   Sweet and Spacious Brooklyn Loft   
11   Maison des Sirenes1,bohemian, luminous apartment   
12                               Midtown Pied-a-terre   
13                   Modern 1 BR / NYC / East Village   
14                Spacious 1 bedroom in luxe building   

                                          description  \
0   Beautiful, spacious skylit studio in the heart...   
1   Enjoy 500 s.f. top floor in 1899 brownstone, w...   
2   We welcome you to stay in our lovely 2 br dupl...   
3   Please don’t expect the luxury here just a bas...   
4   Our best guests are seeking a safe, clean, spa...   
5   Beautiful house, gorgeous garden, patio, cozy ...   
6   Comfortable studio apartment with super comfor...   
7   Beautiful peaceful healthy home<br /><br /><b>...   
8   Large, private loft-like room in a spacious 2-...   
9   Parents/grandparents coming to town, or are yo...   
10  A true open-plan loft in a repurposed factory ...   
11  <b>The space</b><br />I am the lucky owner of ...   
12  HELLO.  PLEASE DO NOT HIT "REQUEST TO BOOK". H...   
13  Awesome, spacious & clean 1 bedroom with a coz...   
14  The room is spacious, the neighborhood is safe...   

                                neighborhood_overview      host_name  \
0   Centrally located in the heart of Manhattan ju...       Jennifer   
1   Just the right mix of urban center and local n...    LisaRoxanne   
2                                                 NaN        Rebecca   
3     Theater district, many restaurants around here.       Shunichi   
4   Our neighborhood is full of restaurants and ca...      MaryEllen   
5   Neighborhood is amazing!<br />Best subways to ...         Laurie   
6                                                 NaN  Allen & Irina   
7   Location: Five minutes to Central Park, Museum...            Cyn   
8   - One stop from the East Village, Lower East S...         Joelle   
9   Residential, village-like atmosphere. Lots of ...           Jane   
10  We've lived here for over 10 years and watched...          Chaya   
11                                                NaN       Nathalie   
12  Quiet residential block near many restaurants ...          Tommi   
13  The east village offers a mixture of old schoo...           Dana   
14                                                NaN           Teri   

                        host_location  \
0   New York, New York, United States   
1   New York, New York, United States   
2   Brooklyn, New York, United States   
3   New York, New York, United States   
4   New York, New York, United States   
5   New York, New York, United States   
6   New York, New York, United States   
7   New York, New York, United States   
8   New York, New York, United States   
9   New York, New York, United States   
10  New York, New York, United States   
11  New York, New York, United States   
12  New York, New York, United States   
13  New York, New York, United States   
14  New York, New York, United States   

                                           host_about  host_response_rate  \
0   A New Yorker since 2000! My passion is creatin...                0.80   
1   Laid-back Native New Yorker (formerly bi-coast...                0.09   
2   Rebecca is an artist/designer, and Henoch is i...                1.00   
3   I used to work for a financial industry but no...                1.00   
4   Welcome to family life w

World Happiness


country  year  Life Ladder  Log GDP per capita  Social support  \
0   Afghanistan  2008     3.723590            7.168690        0.450662   
1   Afghanistan  2009     4.401778            7.333790        0.552308   
2   Afghanistan  2010     4.758381            7.386629        0.539075   
3   Afghanistan  2011     3.831719            7.415019        0.521104   
4   Afghanistan  2012     3.782938            7.517126        0.520637   
5   Afghanistan  2013     3.572100            7.503376        0.483552   
6   Afghanistan  2014     3.130896            7.484583        0.525568   
7   Afghanistan  2015     3.982855            7.466215        0.528597   
8   Afghanistan  2016     4.220169            7.461401        0.559072   
9   Afghanistan  2017     2.661718            7.460144        0.490880   
10      Albania  2007     4.634252            9.077325        0.821372   
11      Albania  2009     5.485470            9.161633        0.833047   
12      Albania  2010     5.268937            9.203026        0.733152   
13      Albania  2011     5.867422            9.230898        0.759434   
14      Albania  2012     5.510124            9.246649        0.784502   

    Healthy life expectancy at birth  Freedom to make life choices  \
0                          49.209663                      0.718114   
1                          49.624432                      0.678896   
2                          50.008961                      0.600127   
3                          50.367298                      0.495901   
4                          50.709263                      0.530935   
5                          51.042980                      0.577955   
6                          51.370525                      0.508514   
7                          51.693527                      0.388928   
8                          52.016529                      0.522566   
9                          52.339527                      0.427011   
10                         66.576630                      0.528605   
11                         67.103607                      0.525223   
12                         67.413696                      0.568958   
13                         67.730408                      0.487496   
14                         68.028885                      0.601512   

    Generosity  Perceptions of corruption  Positive affect  Negative affect  \
0     0.181819                   0.881686         0.517637         0.258195   
1     0.203614                   0.850035         0.583926         0.237092   
2     0.137630                   0.706766         0.618265         0.275324   
3     0.175329                   0.731109         0.611387         0.267175   
4     0.247159                   0.775620         0.710385         0.267919   
5     0.074735                   0.823204         0.620585         0.273328   
6     0.118579                   0.871242         0.531691         0.374861   
7     0.094686                   0.880638         0.553553         0.339276   
8     0.057072                   0.793246         0.564953         0.348332   
9    -0.106340                   0.954393         0.496349         0.371326   
10   -0.016183                   0.874700         0.552678         0.246335   
11   -0.162897                   0.863665         0.640024         0.279257   
12   -0.177533                   0.726262         0.647908         0.300060   
13   -0.210193                   0.877003         0.627659         0.256577   
14   -0.174559                   0.847675         0.606636         0.271393   

    Confidence in national government  Democratic Quality  Delivery Quality  \
0                            0.612072           -1.929690         -1.655084   
1                            0.611545           -2.044093         -1.635025   
2                            0.299357           -1.991810         -1.617176   
3                            0.307386           -1.919018         -1.616221   
4                            0.435440     

Book Reviews


Review  Positive Review
0   This was perhaps the best of Johannes Steinhof...             True
1   This very fascinating book is a story written ...             True
2   The four tales in this collection are beautifu...             True
3   The book contained more profanity than I expec...            False
4   We have now entered a second time of deep conc...             True
5   I don't know why it won the National Book Awar...            False
6   The daughter of a prominent Boston doctor is d...            False
7   I was very disapointed in the book.Basicly the...            False
8   I think in retrospect I wasted my time on this...            False
9   I have a hard time understanding what it is th...            False
10  When Tom returns W/ Huck & Jim from thier craz...             True
11  As a young woman I was appauled by Gurian's la...            False
12  Temperance Brennan is an American forensic ant...            False
13  Lovers of Mr. Rochester beware - in this, his ...             True
14  Typical Feinstein ("A Season on the Brink" exc...            False

## Defining my ML Problem

I have chosen the censusData. I will be predicting the income salary of an individual based on their education level and gender, along with other supporting features. The label for this model is income_binary, which is a binary variable,  greater than 50k(>50k) or less than or equal to 50k(<=50k).

This is a supervised machine learning problem as we are training the model on labeled data that is given to us, and it would be a binary classification based on the variable that has two outcomes of either less than or equal to 50k or greater than 50k. 

The features that I will be using are education, sex, age, hours-per-week,workclass, and occupation. I wanted to add more features so that I have a stronger model to solve the problem of how much education and your gender can influence someone's income. 

This is an important problem because in most cases and historically, women are often paid less than men, so if this is taken a look at in a statistical way through data and modeling, there might be an action for diversity and inclusion in companies that can create value in their work culture. Companies can use this model to address income gaps if there is a larger gap based on gender and education level. 

In [ ]:

#This is the data I am working with:
data = pd.read_csv("data/censusData.csv")

#first I want to preview the data again 
print("column:")
print(data.dtypes)

print("first 50 rows:")
display(data.head(50))

# cheeck missing values because I want to make sure my features that I am using don't have too many missing values 
print("missing values per col")
print(data.isna().sum())

#clean data 
data.dropna(inplace=True)


column:
age               float64
workclass          object
fnlwgt              int64
education          object
education-num       int64
marital-status     object
occupation         object
relationship       object
race               object
sex_selfID         object
capital-gain        int64
capital-loss        int64
hours-per-week    float64
native-country     object
income_binary      object
dtype: object
first 50 rows:


age         workclass  fnlwgt     education  education-num  \
0   39.0         State-gov   77516     Bachelors             13   
1   50.0  Self-emp-not-inc   83311     Bachelors             13   
2   38.0           Private  215646       HS-grad              9   
3   53.0           Private  234721          11th              7   
4   28.0           Private  338409     Bachelors             13   
5   37.0           Private  284582       Masters             14   
6   49.0           Private  160187           9th              5   
7   52.0  Self-emp-not-inc  209642       HS-grad              9   
8   31.0           Private   45781       Masters             14   
9   42.0           Private  159449     Bachelors             13   
10  37.0           Private  280464  Some-college             10   
11  30.0         State-gov  141297     Bachelors             13   
12  23.0           Private  122272     Bachelors             13   
13  32.0           Private  205019    Assoc-acdm             12   
14  40.0           Private  121772     Assoc-voc             11   
15  34.0           Private  245487       7th-8th              4   
16  25.0  Self-emp-not-inc  176756       HS-grad              9   
17  32.0           Private  186824       HS-grad              9   
18  38.0           Private   28887          11th              7   
19  43.0  Self-emp-not-inc  292175       Masters             14   
20  40.0           Private  193524     Doctorate             16   
21  54.0           Private  302146       HS-grad              9   
22  35.0       Federal-gov   76845           9th              5   
23  43.0           Private  117037          11th              7   
24  59.0           Private  109015       HS-grad              9   
25  56.0         Local-gov  216851     Bachelors             13   
26  19.0           Private  168294       HS-grad              9   
27  54.0               NaN  180211  Some-college             10   
28  39.0           Private  367260       HS-grad              9   
29  49.0           Private  193366       HS-grad              9   
30  23.0         Local-gov  190709    Assoc-acdm             12   
31  20.0           Private  266015  Some-college             10   
32  45.0           Private  386940     Bachelors             13   
33  30.0       Federal-gov   59951  Some-college             10   
34  22.0         State-gov  311512  Some-college             10   
35  48.0           Private  242406          11th              7   
36  21.0           Private  197200  Some-college             10   
37  19.0           Private  544091       HS-grad              9   
38  31.0           Private   84154  Some-college             10   
39  48.0  Self-emp-not-inc  265477    Assoc-acdm             12   
40  31.0           Private  507875           9th              5   
41  53.0  Self-emp-not-inc   88506     Bachelors             13   
42  24.0           Private  172987     Bachelors             13   
43  49.0           Private   94638       HS-grad              9   
44  25.0           Private  289980       HS-grad              9   
45  57.0       Federal-gov  337895     Bachelors             13   
46  53.0           Private  144361       HS-grad              9   
47  44.0           Private  128354       Masters             14   
48  41.0         State-gov  101603     Assoc-voc             11   
49  29.0           Private  271466     Assoc-voc             11   

           marital-status         occupation   relationship  \
0           Never-married       Adm-clerical  Not-in-family   
1      Married-civ-spouse    Exec-managerial        Husband   
2                Divorced  Handlers-cleaners  Not-in-family   
3      Married-civ-spouse  Handlers-cleaners        Husband   
4      Married-civ-spouse     Prof-specialty           Wife   
5      Married-civ-spouse    Exec-managerial           Wife   
6   Married-spouse-absent      Other-service  Not-in-family   
7      Married-civ-spouse    Exec-managerial        Husband   
8           Never-ma

missing values per col
age                162
workclass         1836
fnlwgt               0
education            0
education-num        0
marital-status       0
occupation        1843
relationship         0
race                 0
sex_selfID           0
capital-gain         0
capital-loss         0
hours-per-week     325
native-country     583
income_binary        0
dtype: int64


##  Understanding My Data

The next step is to perform exploratory data analysis. Inspect and analyze my data set with my machine learning problem in mind. 

    * addressing missingness, such as replacing missing values with means
    * finding and replacing outliers
    * renaming features and labels
    * finding and replacing outliers
    * performing feature engineering techniques such as one-hot encoding on categorical features
    * selecting appropriate features and removing irrelevant features
    * performing specific data cleaning and preprocessing techniques for an NLP problem
    * addressing class imbalance in your data sample to promote fair AI
    



In [117]:
# doing one-hot encoding: this is for the education and sex 
# Second Try: I am only going to one-hot encode education I don't need to do sex_selfID because I am grouping them diffrently

for col in ['education']:
    if col in data.columns:
        encoded = pd.get_dummies(data[col],prefix=col,drop_first=True)
        data = pd.concat([data.drop(col,axis=1), encoded], axis=1)
print("updated column after one-hot encoding:")

# I wanted to double check if my one-hot encoding worked
print(list(data.columns))

updated column after one-hot encoding:
['age', 'workclass', 'fnlwgt', 'education-num', 'marital-status', 'occupation', 'relationship', 'race', 'sex_selfID', 'capital-gain', 'capital-loss', 'hours-per-week', 'native-country', 'income_binary', 'education_11th', 'education_12th', 'education_1st-4th', 'education_5th-6th', 'education_7th-8th', 'education_9th', 'education_Assoc-acdm', 'education_Assoc-voc', 'education_Bachelors', 'education_Doctorate', 'education_HS-grad', 'education_Masters', 'education_Preschool', 'education_Prof-school', 'education_Some-college']


##  Project Plan





I was able to inspect my data and handle missing values in my previous steps by analyzing and inspecting my data. With this, I finalized my feature list for modeling. I am keeping my two main features: education and sex_selfID.  I did one-hot encoding on these with specific education levels such as education_Bachlors, education_Doctorate, education_Masters etc, along with a binary gender column, for instance, sex_selfID_Non-female, to form my set. I might add more features, such as age and hours-per-week, into my model, but I will first be working with these. 

For the preparation of the data modeling, I dropped rows for missing values, so my data is clean. As mentioned, one-hot encoding was used to convert my categorical columns to numerical as this is best for logistic regression problmes. 

I will be using logistic regression as the model that is best suited for this binary classification problem and works well with one-hot encoding categorical values. 

I will split my dataset into training and testing with train_test_split, and train my logistic regression model on the training data. I will then evaluate my model's performance using accuracy, precision, and recall. 

Based on that if my model accuracy and performance are not good, I will consider adding more features, finding the best weights, or test other model such as Decsion Trees. I might also consider adding fewer features because overfitting can occur with too many features. 


## Part 5: Implement My Project Plan



In [133]:
# YOUR CODE HERE
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, precision_score, recall_score
import matplotlib.pyplot as plt

In [143]:
# YOUR CODE HERE

# One-hot encoded education features, orginally also did sex but took it off because it was not neccsiary 

features = ['education_Bachelors', 'education_Doctorate','education_Masters', 'education_Prof-school','education_Some-college']

# I did not want income_binary in text form I wanted to convert it to binary numbers 0 for <=50k and 1 for >=50k
# this is because I am doing binary classification for my logistic regression model and this was causing me alot of errors before I converted.
if data['income_binary'].dtype == 'object':
    data['income_binary'] = data['income_binary'].map({'<=50K': 0, '>50K': 1})

# to check that data is being grouped properly for both classes 
print(data[['sex_selfID', 'income_binary']].value_counts(dropna=False))

# This is where I came in with my second attempt to train my model because I wanted two groups one for Female and one for Non-Female
#separate models for each group both using logistic regression
gender_groups = {'Female': data[data['sex_selfID'] == 'Female'],'Non-Female': data[data['sex_selfID'] == 'Non-Female']}

#this loops through each gender group 
for label, group_data in gender_groups.items():
    group_data = group_data.dropna(subset=['income_binary'])  # drop rows with missing values because this can cause an error in the model
    
#I had trial and error with this: this is so if there is no data left after removing values it will skip the group
    if group_data.shape[0] == 0:
        print(f"\n{label.upper()} MODEL: skip")
        continue
#input and target value
    X = group_data[features]
    y = group_data['income_binary']
# in case everyon has the same income we can't train the model
    if len(y.unique()) < 2:
        print(f"\n{label.upper()} MODEL: skip")
        continue

    # Train-test split
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=80)

    # train model: logistic regression 
    model = LogisticRegression(max_iter=300)
    model.fit(X_train, y_train)

    # Predict
    y_pred = model.predict(X_test)

    # Evaluate and print for the model
    print(f"\n{label.upper()} MODEL:")
    print("Accuracy:", accuracy_score(y_test, y_pred))
    print("Precision:", precision_score(y_test, y_pred, average='binary'))
    print("Recall:", recall_score(y_test, y_pred, average='binary'))


sex_selfID  income_binary
Non-Female  0                13774
Female      0                 8541
Non-Female  1                 6301
Female      1                 1100
dtype: int64

FEMALE MODEL:
Accuracy: 0.8989113530326595
Precision: 0.5882352941176471
Recall: 0.050505050505050504

NON-FEMALE MODEL:
Accuracy: 0.723785803237858
Precision: 0.5809954751131222
Recall: 0.4984472049689441


***Analyses & Results:***

***The first result for my model has an accuracy of 78.29%, precision of 60.82% and Recall of 42.5%***

*I used 6 one-hot encoded features that represent education levles and also includes gender the features are educaiton_Bachelors, education_Doctorate, education_masters, education_Prof-school, education_Some-college, sex_selfID_Non-Female. 

*I have my target value of income_binary which had numberic values of less then 50k which is 0 and 50k as 1 which helps with binary classification. My data was split into 80/20 and I used the logisctic regression model. 

*The model correctly predicted income class most of the time with 78.3% and with precision 60.8% the model was correct aobut income being greater then 50k.
    *This is where I came into my first obstical that the feature sex_slefID_Non_Female is not being used properly because I wanted to see if someons gender plus their education equalivates to a certain income of either higher then 50k or less than 50k. Because my machine  learning problem is predicting the income salary of an individual based on their education level and gender. My goal is to analyze whether someones gender, along with their education level could be used to predict income, however the current model does not provide clear insights to gender-based income.
    
*What I will do now: Inspect my features, add or remove features, and maybe seperating my data into smaller groups and train and evaluate on seprate models for each group.



-----------------------------------------------------------------------------------------------------


***Second results after seprating the dataset into two gender groups Female and Non-Female and training seprate models***

*My target variable was income_binary , based on this 13,774 non-female make more then 50k, where as there are 8,541 females who make more then 50k. For making less then 50k there are 6,301 non-females and 1,100 females who make less then 50k. High income earner for females are lower. 

*The models performace for the Female Model: Accuracy: 89.89%, Precision:58.82%, Recall 5.05%
*The models performace for the Non-Female Model: Accuracy: 72.38%, Precision: 58.10%, Recall 49.84%

*This shows that the female model has higher accuracy but low recall because it missed alot of the high income female earners. Non-female model performed with better balance of precision and recall which suggests that the data may generalize better for non-female individuals. This noticable income gap and the model performace differances shows the bias and imblance in the dateset espically for females who have a higher income. Another analysis is that there could be better data collection for feature representation for higher income females. 